src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationOptions.cs

```diff
@@ -16,11 +17,7 @@ namespace AspNet.Security.OAuth.Alipay
     /// </summary>
     public class AlipayAuthenticationOptions : OAuthOptions
     {
-        /// <summary>
-        /// Gets or sets a value indicating whether [validate signature].
-        /// </summary>
-        /// <value><c>true</c> if [validate signature]; otherwise, <c>false</c>.</value>
-        public bool ValidateSignature { get; set; } = true;
+        public DateTime? Timestamp { get; set; }
```

Comment:
Revert this part of the change. The `IClock` interface is provided for testability - use that in the tests to "freeze" the clock to the time `new DateTime(2019, 12, 14, 22, 22, 22)` instead.

In fact, in doing that you can also just remove the option to validate the signature entirely and always do it. The tests can freeze the clock so they pass, and users then aren't given an option to make it easy to make the provider less secure.

Reply:
You should be able to do this by adding the following:

```csharp
protected internal override void RegisterAuthentication(AuthenticationBuilder builder)
{
    builder.Services.AddSingleton<ISystemClock, FixedClock>();
    builder.AddAlipay(options =>
    {
        ConfigureDefaults(builder, options);
        options.ClientSecret = "{the key}";
    });
}

private sealed class FixedClock : ISystemClock
{
    public DateTimeOffset UtcNow => new DateTime(2019, 12, 14, 22, 22, 22, DateTimeKind.Utc);
}
```

Reply:
Actually, I think it'll need to be registered before `AddAuthentication()` based on [this code](https://github.com/aspnet/AspNetCore/blob/62351067ff4c1401556725b401478e648b66acdc/src/Security/Authentication/Core/src/AuthenticationServiceCollectionExtensions.cs#L26) that wires up the default `ISystemClock`.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/a865eedad1ad0717fca85c42086a2cd3a34c7d8f/test/AspNet.Security.OAuth.Providers.Tests/Infrastructure/ApplicationFactory.cs#L77

Reply:
Thanks for your help. All Tests Pass.